src/functions/gradients.jl

```diff
@@ -1,3 +1,180 @@
+    T::AbstractVector{<:AbstractFloat},
+    λ::Float64,
+    d::AbstractVector{P}
+) where {P}
+    gradB = _∇acceleration_bezier(M, B, degrees, T)
+    Bt = get_bezier_segments(M, B, degrees, :differentiable )
+    # add start and end data grad
+    # include data term
+    for k=1:length(Bt)
+        gradB[k].pts[1] .+= λ*∇distance(M, d[k], Bt[k].pts[1])
+        if k > 1
+            gradB[k-1].pts[end] .+= λ*∇distance(M, d[k], Bt[k].pts[1])
+        end
+    end
+    gradB[end].pts[end] .+= λ*∇distance(M, d[end], Bt[end].pts[end])
+    return get_bezier_points(M, gradB, :differentiable)
+end
+
+# common helper for the two acceleration grads
+function _∇acceleration_bezier(
+    M::Manifold,
+    B::AbstractVector{P},
+    degrees::AbstractVector{Int},
+    T::AbstractVector{Float64}
+) where {P}
+    Bt = get_bezier_segments(M, B, degrees, :differentiable )
+    n = length(T)
+    m = length(Bt)
+    p = de_casteljau(M,Bt,T)
+    center = p
+    forward = p[ [1, 3:n..., n] ]
+    backward = p[ [1,1:(n-2)..., n] ]
+    mid = mid_point.(Ref(M), backward, forward)
+    # where the point of interest appears...
+    dt = ( max(T...) - min(T...) )/(n-1)
+    inner =  - 2/((dt)^3) .* log.(Ref(M),mid,center)
+    asForward = adjoint_differential_geodesic_startpoint.(Ref(M),forward,backward, Ref(0.5), inner)
+    asCenter = - 2/((dt)^3)*log.(Ref(M),center,mid)
+    asBackward = adjoint_differential_geodesic_endpoint.(Ref(M), forward, backward, Ref(0.5), inner )
+    # effect of these to the centrol points is the preliminary gradient
+    ∇B = [
+        BezierSegment(a.pts .+ b.pts .+ c.pts)
+        for (a,b,c) in zip(
+            adjoint_differential_bezier_control(M, Bt, T[ [1,3:n...,n]],asForward),
+            adjoint_differential_bezier_control(M, Bt, T, asCenter),
+            adjoint_differential_bezier_control(M, Bt, T[ [1,1:(n-2)...,n] ], asBackward)
+        )
+    ]
+    for k=1:length(Bt)-1 # add both effects of left and right segments
+        X = ∇B[k+1].pts[1] + ∇B[k].pts[end]
+        ∇B[k].pts[end] .= X
+        ∇B[k+1].pts[1] .= X
+    end
+    # include c0 & C1 condition
+    for k=length(Bt):-1:2
+        m = length(Bt[k].pts)
+        # updates b-
+        X1 = ∇B[k-1].pts[end-1] .+ adjoint_differential_geodesic_startpoint(
+            M,
+            Bt[k-1].pts[end-1],
+            Bt[k].pts[1],
+            2.,
+            ∇B[k].pts[2]
+        )
+        # update b+ - though removed in reduced form
+        X2 = ∇B[k].pts[2] .+ adjoint_differential_geodesic_startpoint(
+            M,
+            Bt[k].pts[2],
+            Bt[k].pts[1],
+            2.,
+            ∇B[k-1].pts[end-1]
+        )
+        # update p - effect from left and right segment as well as from c1 cond
+        X3 = ∇B[k].pts[1] .+ adjoint_differential_geodesic_endpoint(
+            M,
+            Bt[k-1].pts[m-1],
+            Bt[k].pts[1],
+            2.,
+            ∇B[k].pts[2]
+        )
+        # store
+        ∇B[k-1].pts[end-1] .= X1
```

Comment:
You could probably just do `∇B[k-1].pts[end-1] .+= adjoint_differential_geodesic_startpoint(`?

Reply:
no since the three terms are mixed, so I need the `∇B[k-1].pts[end-1]` in the three computations before (as well as others), i.e. for `X3`, so I think here I need the three allocations for X1, X2 & X3.

Reply:
Right, it's not that easy to improve.